drivers/entropy/entropy_nrf5.c

```diff
@@ -87,11 +88,59 @@ struct entropy_nrf5_dev_data {
 #define DEV_DATA(dev) \
 	((struct entropy_nrf5_dev_data *)(dev)->driver_data)
 
+static void hw_start(struct entropy_nrf5_dev_data *dev_data)
+{
+	unsigned int key;
+
+	__ASSERT_NO_MSG(dev_data);
+
+	key = irq_lock();
+
+	__ASSERT_NO_MSG((dev_data->hw_users + 1) != 0u);
+	if (dev_data->hw_users++ == 0)
+		nrf_rng_task_trigger(NRF_RNG_TASK_START);
+
+	irq_unlock(key);
+}
+
+static void hw_stop(struct entropy_nrf5_dev_data *dev_data)
+{
+	unsigned int key;
+
+	__ASSERT_NO_MSG(dev_data);
+
+	key = irq_lock();
+
+	if (dev_data->hw_users != 0)
+		if (--dev_data->hw_users == 0)
+			nrf_rng_task_trigger(NRF_RNG_TASK_STOP);
+
+	irq_unlock(key);
+}
+
+static int random_byte_get(void)
+{
+	int retval = -EAGAIN;
+	unsigned int key;
+
+	key = irq_lock();
+
+	if (nrf_rng_event_get(NRF_RNG_EVENT_VALRDY)) {
+		retval = nrf_rng_random_value_get();
+		nrf_rng_event_clear(NRF_RNG_EVENT_VALRDY);
+	}
+
+	irq_unlock(key);
+
+	return retval;
+}
+
 #pragma GCC push_options
 #if defined(CONFIG_BT_CTLR_FAST_ENC)
 #pragma GCC optimize ("Ofast")
 #endif
-static inline u8_t get(struct rand *rng, u8_t octets, u8_t *rand)
+static u8_t get(struct entropy_nrf5_dev_data *dev_data,
```

Comment:
Facts:
- After the change, the function has not more that 4 arguments, which means that all arguments are passed in registers.
- The dev_data is already available in caller (which means that it is already loaded to register), because pointer to `rng` is extracted from this structure.

So the penalty will be:
- 0 cycles, if compiler will keep `dev_data` in `r0`.
- 1 cycle, if compiler will use any other register to store `dev_data`.

1 cycle is:
 - 62.5ns on nRF51.
 - 15.625ns on nRF52.

Do you think that 1 cycle penalty could affect some time-critical code in Zephyr?
If so, I think that we have a much bigger problem...

Reply:
Radio ISR profiling data:
l -- ISR entry latency (current, min, and max)
t -- ISR CPU time (current, min and max)

before this PR:
[2018-08-24 14:18:33.596] [bt] [INF] encode_control: l: 8, 7, 8; t: 60, 47, 87.
[2018-08-24 14:18:33.626] [bt] [INF] encode_control: l: 7, 7, 8; t: 131, 47, 131.
[2018-08-24 14:18:33.655] [bt] [INF] encode_control: l: 7, 7, 8; t: 78, 47, 131.
[2018-08-24 14:18:33.686] [bt] [INF] encode_control: l: 7, 7, 8; t: 113, 47, 131.
[2018-08-24 14:18:33.715] [bt] [INF] encode_control: l: 7, 7, 8; t: 86, 47, 131.

after this PR:
[2018-08-24 14:16:32.368] [bt] [INF] encode_control: l: 7, 7, 8; t: 60, 44, 87.
[2018-08-24 14:16:32.398] [bt] [INF] encode_control: l: 8, 7, 8; t: 136, 44, 136.
[2018-08-24 14:16:32.427] [bt] [INF] encode_control: l: 7, 7, 8; t: 78, 44, 136.
[2018-08-24 14:16:32.458] [bt] [INF] encode_control: l: 7, 7, 8; t: 113, 44, 136.
[2018-08-24 14:16:32.487] [bt] [INF] encode_control: l: 7, 7, 8; t: 87, 44, 136.

There is an increase of 5 us (~ 4%) in the Radio ISR on nRF51, and 136 us is very close to the limit, and any added latencies from the OS will cause an assert.

Reply:
I think that this is not caused by single cycle penalty mentioned here, but rather something more complex (`irq_lock()`/`irq_unlock()` around `get()`, `entropy_get_entropy_isr()` wrapper, etc.).

Could you please tell me how to run the test you shown? I would like to trace down source of the delay, as 5us increase seems to be a bit too big comparing to code changes (i expected 1-2us).

Reply:
I am refering to LDR instructions needed to load Rn registers to pass parameters, LDR is 2 cycles; then there is indirect memory access inside the function, which is another 2 cycles of LDR. There will be many MOV, CMP and B instructions. This may still seem minute, but when the function is called many times, these cycles add up quickly.

build the samples/bluetooth/peripheral application with CONFIG_BT_CTLR_PROFILE_ISR=y, CONFIG_BT_CTLR_ADVANCED_FEATURES=y, and CONFIG_BT_CTLR_FAST_ENC=y.
Then use a phone to bond, i.e. have encryption setup procedure performed. In the UART you will get the Radio ISR profiling printed.

Alternatively, zephyr.lst file can be used to compare the changes in resultant assembly instructions increased by changes.